source/solvers/mf_navier_stokes.cc

```diff
@@ -44,6 +45,209 @@
 
 #include <deal.II/numerics/vector_tools.h>
 
+template <typename VectorType>
+class PreconditionBase
+{
+public:
+  virtual void
+  vmult(VectorType &dst, const VectorType &src) const = 0;
+};
+
+template <typename VectorType>
+class MyDiagonalMatrix : public PreconditionBase<VectorType>
+{
+public:
+  MyDiagonalMatrix(const VectorType &diagonal)
+    : diagonal_matrix(diagonal)
+  {}
+
+  void
+  vmult(VectorType &dst, const VectorType &src) const override
+  {
+    diagonal_matrix.vmult(dst, src);
+  }
+
+private:
+  DiagonalMatrix<VectorType> diagonal_matrix;
+};
+
+template <typename VectorType>
+class PreconditionASM : public PreconditionBase<VectorType>
+{
+private:
+  enum class WeightingType
+  {
+    none,
+    left,
+    right,
+    symm
+  };
+
+public:
+  using Number = typename VectorType::value_type;
+
+  PreconditionASM()
+    : weighting_type(WeightingType::left)
+  {}
```

Comment:
So right now this is still a prototype class I presume right (because the weighting type of hardcoded).

Reply:
> So right now this is still a prototype class I presume right (because the weighting type of hardcoded).

Yes its still a prototype. I need to make some clean up; but the version here should be enough to conduct some first experiments.

For the weighting type, one probably does not need any parameter, since the others don't really work in the NS context (I wrote the original implementation for a Poisson operator where the other variants worked better).

Reply:
Perfect, good for me.